src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -84,14 +84,29 @@ master_create_empty_shard(PG_FUNCTION_ARGS)
 	char storageType = SHARD_STORAGE_TABLE;
 
 	Oid relationId = ResolveRelationId(relationNameText);
+	char relationKind = get_rel_relkind(relationId);
 	char *relationOwner = TableOwner(relationId);
 
 	EnsureTablePermissions(relationId, ACL_INSERT);
 	CheckDistributedTable(relationId);
 
-	if (CStoreTable(relationId))
+	/*
+	 * We check whether the table is a foreign table or not. If it is, we set
+	 * storage type as foreign also. Only exception is if foreign table is a
+	 * foreign cstore table, in this case we set storage type as columnar.
+	 *
+	 * i.e. While setting storage type, columnar has priority over foreign.
+	 */
+	if (relationKind == RELKIND_FOREIGN_TABLE)
 	{
-		storageType = SHARD_STORAGE_COLUMNAR;
+		if (CStoreTable(relationId))
+		{
+			storageType = SHARD_STORAGE_COLUMNAR;
+		}
+		else
+		{
+			storageType = SHARD_STORAGE_FOREIGN;
+		}
 	}
```

Comment:
This code assumes that relationKind can be either a foreign table or normal table.

A similar check is also done in master_get_table_metadata(): 
```
	if (relationType == RELKIND_RELATION)
	{
		storageType = SHARD_STORAGE_TABLE;
	}
	else if (relationType == RELKIND_FOREIGN_TABLE)
	{
		bool cstoreTable = CStoreTable(relationId);
		if (cstoreTable)
		{
			storageType = SHARD_STORAGE_COLUMNAR;
		}
		else
		{
			storageType = SHARD_STORAGE_FOREIGN;
		}
	}
```

Possibilities for get_rel_relkind output can be found in pg_class.h:
```
RELKIND_RELATION		
RELKIND_INDEX			
RELKIND_SEQUENCE		
RELKIND_TOASTVALUE	
RELKIND_VIEW			
RELKIND_COMPOSITE_TYPE
RELKIND_FOREIGN_TABLE 
RELKIND_MATVIEW	
```	

In our case, can relationType be something other than RELKIND_RELATION and RELKIND_FOREIGN_TABLE (for example a view or something  If so, what should we do then?